numba/core/caching.py

```diff
@@ -318,6 +318,42 @@ def from_function(cls, py_func, py_file):
         return self
 
 
+class _StringSrcCacheLocator(_UserWideCacheLocator):
+
+    def __init__(self, py_func, py_file):
+        self._identifier = self._hash(py_func)
+        super().__init__(py_func, py_file)
+
+    def get_source_stamp(self):
+        return self._identifier
+
+    def get_disambiguator(self):
+        # have to be quite fussy about this as there's very little aliasing
+        # protection with strings (no module or source to check).
+        return self._identifier
+
+    @classmethod
+    def _hash(cls, py_func):
+        buf = []
+        for x in py_func.__code__.co_consts:
+            buf.append(str(x))
+        const_bytes = (''.join(buf)).encode('UTF-8')
+        data = py_func.__code__.co_code + const_bytes
```

Comment:
This won't pick up changes in function name, function arguments, etc. Maybe consider all states in https://github.com/python/cpython/blob/66d3b589c44fcbcf9afe1e442d9beac3bd8bcd34/Objects/codeobject.c#L989-L1012

Might also need the function type signature as this does not store the value of default args.

Reply:
thanks, I've been considering https://github.com/numba/numba/pull/6406#issuecomment-713861806 and generally poking at the contents of `codeobject` to see what can/cannot be easily worked out.

Reply:
My main concern here is freevars and globals, and wiring through enough data to achieve accurate and correct behaviour.

Reply:
I am just not sure if we can handle freevars and globals as is. I am open to the idea of making cacheable functions handle freevars and globals differently than non-cacheable functions. I'm thinking making them linked at load time.